src/rust/pants_ng/options/src/pants_invocation.rs

```diff
@@ -0,0 +1,220 @@
+    LazyLock::new(|| Regex::new(r"^[A-Za-z][0-9A-Za-z_\-]*$").unwrap());
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Args {
+    pub(crate) arg_strings: Vec<String>,
+}
+
+impl Args {
+    // Create an Args instance with the provided args, which must *not* include the
+    // argv[0] process name.
+    pub fn new<I: IntoIterator<Item = String>>(arg_strs: I) -> Self {
+        Self {
+            arg_strings: arg_strs.into_iter().collect(),
+        }
+    }
+
+    pub fn argv() -> Self {
+        let mut args = env::args().collect::<Vec<_>>().into_iter();
+        args.next(); // Consume the process name (argv[0]).
+        // TODO: In Pants's own integration tests we may invoke Pants in a subprocess via
+        //  `python -m pants` or `python path/to/__main__.py` or similar. So
+        //  skipping argv[0] may not be sufficient to get just the set of args to split.
+        //  In practice our tests pass despite these extra args being interpreted as specs
+        //  or goals, but that is skating on thin ice.
+        Self::new(args.collect::<Vec<_>>())
+    }
+}
+
+// Represents a single cli flag used to set an option value, e.g., `--foo`, or `--bar=baz`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Flag {
+    pub key: String, // The flag name, without the `--` prefix, up to the `=` if any.
+    pub value: Option<String>, // The value after the `=`, if any.
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct SubCommand {
+    pub name: String,
+    pub flags: Vec<Flag>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Command {
+    pub name: String,
+    pub flags: Vec<Flag>,
+    pub subcommand: Option<SubCommand>,
+}
+
+// The details of a Pants invocation command.
+//
+// An invocation has the form:
+//
+// pants --global-flags \
+//   cmd1 --cmd1-flags (subcmd1 --subcmd1-flags)? \
+//   + cmd2 --cmd2-flags (subcmd2 --subcmd2-flags)? \
+//   ... \
+//   + cmdN --cmdN-flags (subcmdN --subcmdN-flags)? \
+//   path/to/spec1 path/to/spec2 ... path/to/specM \
+//   (-- passthru-args)?
+//
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct PantsInvocation {
+    pub global_flags: Vec<Flag>,
+    pub commands: Vec<Command>,
+    pub specs: Vec<String>, // Non-flag arguments, typically file paths or globs.
+    pub passthru: Option<Vec<String>>, // Any remaining args specified after a -- separator.
+}
+
+impl PantsInvocation {
+    pub fn empty() -> Self {
+        Self {
+            global_flags: vec![],
+            commands: vec![],
+            specs: vec![],
+            passthru: None,
+        }
+    }
+
+    pub fn from_args(args: Args) -> Result<Self, String> {
+        #[inline]
+        fn is_flag(s: &str) -> bool {
+            s.starts_with("--") && s.len() > 2
+        }
+
+        #[inline]
+        fn is_name(s: &str) -> bool {
+            NAME_RE.is_match(s)
+        }
+
+        #[inline]
+        fn is_spec(s: &str) -> bool {
+            s.contains(path::MAIN_SEPARATOR)
```

Comment:
I don't know rust. When we support Windows at some point, will this be `\` on Windows instead of `/`? Or will it take both? Or does it always use the *nix `/`?

Reply:
I think not at first. Maybe if there is demand for it. But we are a thousand other steps away from even evaluating that demand...